src/transformStateWithClones.js

```diff
@@ -6,8 +6,43 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+
+  for (const action of actions) {
+    const { type = null, extraData = {}, keysToRemove = [] } = action;
+
+    let newObj;
+
+    if (states.length === 0) {
+      newObj = { ...state };
+    } else {
+      newObj = { ...states[states.length - 1] };
+    }
+
+    switch (type) {
+      case 'addProperties':
+        newObj = Object.assign(newObj, extraData);
```

Comment:
The use of `Object.assign` is unnecessary here since you are assigning properties to `newObj` which is already a copy of the previous state. You can directly add properties to `newObj` using the spread operator or by assigning them directly.